### PR TITLE
Cutmix callback 

### PR DESCRIPTION
**Pull request by LightnessOfBeing**
## Description

Implementation of [Cutmix](https://arxiv.org/pdf/1905.04899.pdf) augmentation based on [this](https://www.kaggle.com/c/bengaliai-cv19/discussion/126504) post.

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have read I need to click 'Login as guest' to see Teamcity build logs
- [x] I have checked the code-style using `make check-codestyle`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.

### CODE REVIEW

**Comment by Scitator** [comment]
@LightnessOfBeing Thanks for PR, don't you mind to
- rename `cutmixcallback.py` to `cutmix_callback.py`
- rename `metricaggregatorcallback.py` to `metric_aggregator_callback.py`
- fix the codestyle issue, please check https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md for more info

**Comment by LightnessOfBeing** [comment]
Running `make codestyle` gives me:

```
pre-commit run
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/kirill/.cache/pre-commit/patch1580904366.
yapf.................................................(no files to check)Skipped
seed isort known_third_party.............................................Passed
isort................................................(no files to check)Skipped
Check Yaml...........................................(no files to check)Skipped
Trim Trailing Whitespace.............................(no files to check)Skipped
Flake8...............................................(no files to check)Skipped
[INFO] Restored changes from /home/kirill/.cache/pre-commit/patch1580904366.
```

Running `make check-codestyle` gives me:

```
bash ./bin/_check_codestyle.sh -s
parse -s flag
isort: `isort -rc --check-only --settings-path ./setup.cfg`

Skipped 99 files
flake8: `bash ./bin/flake8.sh`

# this stops git rev-parse from failing if we run this from the .git directory
builtin cd "$(dirname "${BASH_SOURCE:-$0}")"

ROOT="$(git rev-parse --show-toplevel)"
builtin cd "$ROOT" || exit 1

git remote add 'upstream' 'https://github.com/catalyst-team/catalyst' || true
fatal: remote upstream already exists.

# Only fetch master since that's the branch we're diffing against.
git fetch upstream master
From https://github.com/catalyst-team/catalyst
 * branch            master     -> FETCH_HEAD

MERGEBASE="$(git merge-base upstream/master HEAD)"
if ! git diff --diff-filter=ACM --quiet --exit-code "$MERGEBASE" -- '*.py' &>/dev/null; then
    flake8 "$@" $(git diff --name-only --diff-filter=AM "$MERGEBASE" -- '*.py')
fi
0
flake8 (warnings): `flake8`
./catalyst/contrib/models/sequential.py:31:5: C901 'SequentialNet.__init__' is too complex (12)
./catalyst/contrib/nn/optimizers/radam.py:26:5: C901 'RAdam.step' is too complex (11)
./catalyst/contrib/nn/optimizers/ralamb.py:43:5: C901 'Ralamb.step' is too complex (13)
./catalyst/core/experiment.py:3:1: F401 'collections.OrderedDict' imported but unused
./catalyst/core/experiment.py:6:1: F401 'torch.utils.data.DataLoader' imported but unused
./catalyst/core/experiment.py:6:1: F401 'torch.utils.data.Dataset' imported but unused
./catalyst/core/experiment.py:9:1: F401 '.callback.Callback' imported but unused
./catalyst/core/callbacks/phase.py:2:1: F401 'collections.OrderedDict' imported but unused
./catalyst/dl/experiment/base.py:1:1: F401 'typing.List' imported but unused
./catalyst/dl/experiment/base.py:1:1: F401 'typing.Union' imported but unused
./catalyst/dl/experiment/base.py:2:1: F401 'collections.OrderedDict' imported but unused
./catalyst/dl/experiment/base.py:5:1: F401 'torch.utils.data.DataLoader' imported but unused
./catalyst/dl/experiment/base.py:7:1: F401 'catalyst.dl.Callback' imported but unused
./catalyst/dl/experiment/config.py:14:1: F401 'catalyst.dl.Callback' imported but unused
./catalyst/dl/experiment/config.py:185:5: C901 'ConfigExperiment._get_optimizer' is too complex (13)
./catalyst/dl/experiment/config.py:387:5: C901 'ConfigExperiment.get_loaders' is too complex (14)
./catalyst/dl/experiment/supervised.py:1:1: F401 'collections.OrderedDict' imported but unused
./catalyst/dl/experiment/supervised.py:3:1: F401 'catalyst.dl.Callback' imported but unused
./catalyst/dl/runner/supervised.py:1:1: F401 'typing.List' imported but unused
./catalyst/dl/runner/supervised.py:10:1: F401 'catalyst.dl.Callback' imported but unused
./catalyst/dl/runner/supervised.py:301:5: C901 'SupervisedRunner.trace' is too complex (11)
./catalyst/rl/offpolicy/algorithms/sac.py:8:1: F401 'catalyst.rl.core.AlgorithmSpec' imported but unused
./catalyst/rl/offpolicy/algorithms/td3.py:8:1: F401 'catalyst.rl.core.AlgorithmSpec' imported but unused
./catalyst/rl/scripts/run_samplers.py:162:1: C901 'main' is too complex (12)
./catalyst/utils/config.py:215:1: C901 'parse_config_args' is too complex (12)
./catalyst/utils/torch.py:291:1: C901 'process_components' is too complex (13)
./catalyst/utils/meters/apmeter.py:34:5: C901 'APMeter.add' is too complex (12)
10    C901 'SequentialNet.__init__' is too complex (12)
17    F401 'collections.OrderedDict' imported but unused
27
yapf: `bash ./bin/yapf.sh --all-in-place`

# this stops git rev-parse from failing if we run this from the .git directory
builtin cd "$(dirname "${BASH_SOURCE:-$0}")"

ROOT="$(git rev-parse --show-toplevel)"
builtin cd "$ROOT" || exit 1

git remote add 'upstream' 'https://github.com/catalyst-team/catalyst' || true
fatal: remote upstream already exists.

# Only fetch master since that's the branch we're diffing against.
git fetch upstream master
From https://github.com/catalyst-team/catalyst
 * branch            master     -> FETCH_HEAD

YAPF_FLAGS=(
    '--style' "$ROOT/setup.cfg"
    '--recursive'
    '--parallel'
)

YAPF_EXCLUDES=(
    '--exclude' 'docker/*'
)

# Format specified files
format() {
    yapf --in-place "${YAPF_FLAGS[@]}" -- "$@"
}

# Format files that differ from main branch. Ignores dirs that are not slated
# for autoformat yet.
format_changed() {
    # The `if` guard ensures that the list of filenames is not empty, which
    # could cause yapf to receive 0 positional arguments, making it hang
    # waiting for STDIN.
    #
    # `diff-filter=ACM` and $MERGEBASE is to ensure we only format files that
    # exist on both branches.
    MERGEBASE="$(git merge-base upstream/master HEAD)"

    if ! git diff --diff-filter=ACM --quiet --exit-code "$MERGEBASE" -- '*.py' &>/dev/null; then
        git diff --name-only --diff-filter=ACM "$MERGEBASE" -- '*.py' | xargs -P 5 \
             yapf --in-place "${YAPF_EXCLUDES[@]}" "${YAPF_FLAGS[@]}"
    fi
}

# Format all files, and print the diff to stdout for travis.
format_all() {
    yapf --diff "${YAPF_FLAGS[@]}" "${YAPF_EXCLUDES[@]}" ./**/**/*.py
}

format_all_in_place() {
    yapf --in-place "${YAPF_FLAGS[@]}" "${YAPF_EXCLUDES[@]}" ./**/**/*.py
}

# This flag formats individual files. --files *must* be the first command line
# arg to use this option.
if [[ "$1" == '--files' ]]; then
    format "${@:2}"
    # If `--all` is passed, then any further arguments are ignored and the
    # entire python directory is formatted.
elif [[ "$1" == '--all' ]]; then
    format_all
elif [[ "$1" == '--all-in-place' ]]; then
    format_all_in_place
else
    # Format only the files that changed in last commit.
    format_changed
fi

if ! git diff --quiet &>/dev/null; then
    echo 'Reformatted changed files. Please review and stage the changes.'
    echo 'Files updated:'
    echo

    git --no-pager diff --name-only

    exit 1
fi
Reformatted changed files. Please review and stage the changes.
Files updated:

catalyst/core/runner.py
catalyst/utils/dict.py
catalyst/utils/initialization.py
catalyst/utils/pandas.py
examples/_test_segmentation/dataset.py
examples/mnist_gans/experiment.py
Makefile:4: recipe for target 'check-codestyle' failed
make: *** [check-codestyle] Error 1
```


Do I need to commit and push the updated files in the end, even though I haven't made any changes in them on purpose? The changes seem only be related to formatting.

**Comment by Scitator** [comment]
you need to run `make codestyle` after you add them to git (`git add`), but before you commit them into it (`git commit`)

nevertheless, you can use
```
isort -rc --settings-path ./setup.cfg
bash ./bin/yapf.sh --all-in-place
```
to make the codestyle great again

File: catalyst/core/runner.py [hunk 1]
```diff
@@ -241,9 +242,8 @@ def _run_batch_train_step(self, batch: Mapping[str, Any]):
         self.state.output = self.forward(batch)
 
     @torch.no_grad()
-    def predict_batch(
-        self, batch: Mapping[str, Any], **kwargs
-    ) -> Mapping[str, Any]:
+    def predict_batch(self, batch: Mapping[str, Any],
+                      **kwargs) -> Mapping[str, Any]:
```
Comment (Scitator) [inline catalyst/core/runner.py]: codestyle please :)

File: catalyst/utils/initialization.py [hunk 2]
```diff
@@ -22,9 +22,8 @@ def _nonlinearity2name(nonlinearity):
     return nonlinearity
 
 
-def create_optimal_inner_init(
-    nonlinearity: nn.Module, **kwargs
-) -> Callable[[nn.Module], None]:
+def create_optimal_inner_init(nonlinearity: nn.Module,
+                              **kwargs) -> Callable[[nn.Module], None]:
```
Comment (Scitator) [inline catalyst/utils/initialization.py]: codestyle again :)

File: catalyst/utils/pandas.py [hunk 3]
```diff
@@ -236,9 +240,8 @@ def separate_tags(
     return df_new
 
 
-def get_dataset_labeling(
-    dataframe: pd.DataFrame, tag_column: str
-) -> Dict[str, int]:
+def get_dataset_labeling(dataframe: pd.DataFrame,
+                         tag_column: str) -> Dict[str, int]:
```
Comment (Scitator) [inline catalyst/utils/pandas.py]: codestyle

**Comment by Scitator** [comment]
@bagxi @TezRomacH @lightforever what do you think about our current auto codestyle setting?

**Comment by TezRomacH** [comment]
auto code style works well with `pre-commit run`

File: catalyst/core/runner.py (hunk 1 above)
Comment (LightnessOfBeing) [reply catalyst/core/runner.py]: Well, this one was actually made by `yapf`, i.e `bash ./bin/yapf.sh --all-in-place`

Comment (LightnessOfBeing) [reply catalyst/core/runner.py]: Probably my local codestyle scripts are not up to date?

Comment (LightnessOfBeing) [reply catalyst/core/runner.py]: Actually, `yapf` restores all three your suggestions back to the current revision.

**Comment by lightforever** [comment]
There is a slight inconvenience: to see whether the code satisfies our rules, we need to make a commit.

`pre-commit run` helps to cope with that.

File: catalyst/contrib/dl/callbacks/__init__.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 # flake8: noqa
 
+from .cutmix_callback import CutmixCallback
 from .knn import KNNMetricCallback
-from .metricaggregatorcallback import MetricAggregatorCallback
+from .metric_aggregator_callback import MetricAggregatorCallback
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/__init__.py:5]: btw, speaking about naming, there is a good questions: do we really need callback postfix for .py files?
why not to use just `cutmix.py` and `metric_aggregator.py`?

File: catalyst/contrib/dl/callbacks/cutmix_callback.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        if not self.is_needed:
+            return super()._compute_loss(state, criterion)
+
+        pred = state.output[self.output_key]
+        y_a = state.input[self.input_key]
+        y_b = state.input[self.input_key][self.index]
+        loss = self.lam * criterion(pred, y_a) + \
+            (1 - self.lam) * criterion(pred, y_b)
+        return loss
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/cutmix_callback.py]: please place all protected methods before (upper) public methods due to our codestyle guidelines

File: catalyst/core/runner.py (hunk 1 above)
Comment (Scitator) [reply catalyst/core/runner.py]: please run `pip install -r ./requirements/requirements-dev.txt` for correct yapf version

**Comment by bagxi** [comment]
> @bagxi @TezRomacH @lightforever what do you think about our current auto codestyle setting?

In my opinion, the easiest way to follow the catalyst's code style and avoid related issues is to read [contribution best practices](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md#contribution-best-practices) before starting working on PRs.
Another great option is to really run `make check-codestyle` command rather than only marking in the check-list that this step was done.

**Pull request merged**